js/src/Ice/ConnectionI.js

```diff
@@ -415,14 +415,26 @@ export class ConnectionI {
         let info = null;
         let response = null;
         try {
-            if ((operation & SocketOperation.Write) !== 0 && this._writeStream.buffer.remaining > 0) {
-                DEV: console.assert(this._sendStreams.length > 0);
-                const currentMessage = this._sendStreams[0];
-                if (!this.write(this._writeStream.buffer, () => (currentMessage.isSent = true))) {
-                    DEV: console.assert(!this._writeStream.isEmpty());
-                    return;
+            if ((operation & SocketOperation.Write) !== 0) {
+                if (this._writeStream.buffer.remaining > 0) {
+                    DEV: console.assert(this._sendStreams.length > 0);
+                    const completedSynchronously = this.write(this._writeStream.buffer);
+                    if (this._writeStream.buffer.remaining === 0) {
+                        // If the write call consumed the complete buffer, we assume the message is sent now for
+                        // at-most-once semantics.
+                        this._sendStreams[0].isSent = true;
+                    }
+
+                    if (!completedSynchronously) {
+                        DEV: console.assert(!this._writeStream.isEmpty());
+                        return;
+                    }
+                    DEV: console.assert(this._writeStream.buffer.remaining === 0);
+                } else if (this._transceiver instanceof IdleTimeoutTransceiverDecorator) {
+                    // The writing of the current message completed, schedule a heartbeat in case the connection has
+                    // nothing more to write.
+                    this._transceiver.scheduleHeartbeat();
```

Comment:
This ensure that a heartbeat is scheduled even if write completes asynchronously. In this case Transceiver.write is not called and the decorator would not schedule one.